src/main/java/org/testng/internal/DynamicGraph.java

```diff
@@ -69,75 +48,105 @@ public boolean addNode(T node) {
    * @param to - Represents the edge on which another edge depends upon.
    */
   public void addEdge(int weight, T from, T to) {
-    addEdges(Collections.singletonList(new Edge<>(weight, from, to)));
+    Integer reversedEdgeWeight = findReversedEdge(from, to);
+    if (reversedEdgeWeight != null && reversedEdgeWeight == weight) {
+      throw new IllegalStateException("Circular dependency: " + from + " <-> " + to);
+    }
+
+    addEdgeToMap(m_incomingEdges, to, from, weight);
+    addEdgeToMap(m_outgoingEdges, from, to, weight);
+  }
+
+
+  private void addEdgeToMap(Map<T, Map<T, Integer>> map, T n1, T n2, int weight) {
+    Map<T, Integer> edges = map.get(n1);
+    if (edges == null) {
+      edges = new HashMap<>();
+      map.put(n1, edges);
+    }
+
+    Integer existingWeight = edges.get(n2);
+    edges.put(n2, Math.max(weight, existingWeight != null ? existingWeight : Integer.MIN_VALUE));
   }
 
   /**
    * Add an edge between two nodes.
    */
-  public void addEdge(int weight, T from, T... tos) {
-    addEdge(weight, from, Arrays.asList(tos));
+  @SafeVarargs
+  public final void addEdges(int weight, T from, T... tos) {
+    for (T to : tos) {
+      addEdge(weight, from, to);
+    }
   }
 
   /**
    * Add an edge between two nodes.
    */
-  public void addEdge(int weight, T from, Iterable<T> tos) {
-    List<Edge<T>> edges = Lists.newArrayList();
+  public void addEdges(int weight, T from, Iterable<T> tos) {
     for (T to : tos) {
-      edges.add(new Edge<>(weight, from, to));
+      addEdge(weight, from, to);
     }
-    addEdges(edges);
   }
 
-  private void addEdges(List<Edge<T>> edges) {
-    for (Edge<T> edge : edges) {
-      Edge<T> existingEdge = getNode(m_edges, edge);
-      if (existingEdge != null && existingEdge.weight == edge.weight) {
-        throw new IllegalStateException("Circular dependency: " + edge.from + " <-> " + edge.to);
-      }
-      if ((existingEdge == null) || (existingEdge.weight < edge.weight)) {
-        m_edges.put(edge.from, edge);
+  /**
+   * Remove a node from the graph and all associated edges.
+   *
+   * @param node Node to remove.
+   */
+  private void removeNode(T node) {
+    Map<T, Integer> incomingEdges = m_incomingEdges.get(node);
+    Map<T, Integer> outgoingEdges = m_outgoingEdges.get(node);
+
+    m_incomingEdges.remove(node);
+    m_outgoingEdges.remove(node);
+
+    if (incomingEdges != null) {
+      for (T from : incomingEdges.keySet()) {
+        Map<T, Integer> edges = m_outgoingEdges.get(from);
+        edges.remove(node);
+        if (edges.isEmpty()) {
+          m_outgoingEdges.remove(from);
+        }
       }
-      // else: existingEdge.weight > edge.weight and ignore
-      m_allEdges.put(edge.from, edge);
     }
-  }
 
-  private static <T> Edge<T> getNode(ListMultiMap<T, Edge<T>> edges, Edge<T> edge) {
-    for (Edge<T> e : edges.get(edge.to)) {
-      if (e.to.equals(edge.from)) {
-        return e;
+    if (outgoingEdges != null) {
```

Comment:
Could you add a one line comment on the 2 blocks which explains what you are doing?